keras_cv/models/object_detection/yolo_v8/yolo_v8_detector.py

```diff
@@ -0,0 +1,538 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+
+import tensorflow as tf
+from keras import layers
+from tensorflow import keras
+
+import keras_cv
+from keras_cv import bounding_box
+from keras_cv.models.backbones.backbone_presets import backbone_presets
+from keras_cv.models.backbones.backbone_presets import (
+    backbone_presets_with_weights,
+)
+from keras_cv.models.object_detection import predict_utils
+from keras_cv.models.object_detection.__internal__ import unpack_input
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_detector_presets import (
+    yolo_v8_detector_presets,
+)
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_iou_loss import (
+    YOLOV8IoULoss,
+)
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_label_encoder import (
+    YOLOV8LabelEncoder,
+)
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_layers import (
+    apply_conv_bn,
+)
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_layers import (
+    apply_csp_block,
+)
+from keras_cv.models.task import Task
+from keras_cv.utils.python_utils import classproperty
+from keras_cv.utils.train import get_feature_extractor
+
+
+def get_anchors(
+    image_shape,
+    strides=[8, 16, 32],
+    base_anchors=[0.5, 0.5],
+):
+    """Gets anchor boxes for YOLOV8.
+
+    Returns a tuple of anchor centerpoints and anchor strides.
+    Multiplying the two together will yield the centerpoints in absolute x,y
+    format. The width and height of each anchor box is represented by its
+    corresponding stride.
+    """
+    base_anchors = tf.constant(base_anchors, dtype=tf.float32)
+
+    all_anchors = []
+    all_strides = []
+    for stride in strides:
+        hh_centers = tf.range(start=0, limit=image_shape[0], delta=stride)
+        ww_centers = tf.range(start=0, limit=image_shape[1], delta=stride)
+        ww_grid, hh_grid = tf.meshgrid(ww_centers, hh_centers)
+        grid = tf.cast(
+            tf.reshape(tf.stack([hh_grid, ww_grid], 2), [-1, 1, 2]),
+            tf.float32,
+        )
+        anchors = tf.expand_dims(base_anchors * [stride, stride], 0) + grid
+        anchors = tf.reshape(anchors, [-1, 2])
+        all_anchors.append(anchors)
+        all_strides.append(tf.repeat(stride, anchors.shape[0]))
+
+    all_anchors = tf.cast(tf.concat(all_anchors, axis=0), tf.float32)
+    all_strides = tf.cast(tf.concat(all_strides, axis=0), tf.float32)
+
+    all_anchors = all_anchors / all_strides[:, None]
+
+    # Swap the x and y coordinates of the anchors.
+    all_anchors = tf.concat(
+        [all_anchors[:, 1, tf.newaxis], all_anchors[:, 0, tf.newaxis]], axis=-1
+    )
+    return all_anchors, all_strides
+
+
+def path_aggregation_fpn(features, depth=3, name=None):
```

Comment:
Create a docstring here as well

Reply:
Sadly this diagram is kinda gibberish to me

Reply:
This whole function does not appear to be human readable.

```
for id, ii in enumerate(upsamples[:-1][::-1]):
    ya following?
```

🤔 🤔 🤔

Reply:
I've re-written the entire FPN block to try to make it a bit easier to understand. Is it more understandable now?